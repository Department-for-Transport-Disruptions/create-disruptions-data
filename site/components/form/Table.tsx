import { ReactElement, ReactNode } from "react";

export interface TableStyles {
    width?: string;
}

export interface CellProps {
    value: string | ReactNode;
    styles?: TableStyles;
}

interface TableProps {
    caption?: { text: string; size: "s" | "m" | "l" };
    columns?: string[];
    rows: { header?: string | ReactNode; cells: string[] | ReactNode[] | CellProps[] }[];
}

const isCellProps = (cell: string | ReactNode | CellProps): cell is CellProps => {
    if (cell && typeof cell === "object" && "value" in cell) {
        return true;
    } else {
        return false;
    }
};

const Table = ({ caption, columns = [], rows }: TableProps): ReactElement => {
    return (
        <table className="govuk-table">
            {caption ? (
                <caption className={`govuk-table__caption govuk-table__caption--${caption.size}`}>
                    {caption.text}
                </caption>
            ) : (
                ""
            )}
            <thead className="govuk-table__head">
                <tr className="govuk-table__row">
                    {columns.map((column) => (
                        <th scope="col" className="govuk-table__header align-middle px-2" key={column}>
                            {column}
                        </th>
                    ))}
                </tr>
            </thead>
            <tbody className="govuk-table__body">
                {rows.map((row, index) => (
                    <tr className="govuk-table__row" key={`row-${index}`}>
                        {row.header && (
                            <th scope="row" className="govuk-table__header align-middle">
                                {row.header}
                            </th>
                        )}
                        {row.cells.map((cell, i) => (
                            <td
                                className={`govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl ${
                                    isCellProps(cell) && cell.styles?.width ? cell.styles?.width : ""
                                }`}
                                key={i}
                            >
                                {isCellProps(cell) ? cell.value : cell}
                            </td>
                        ))}
                    </tr>
                ))}
            </tbody>
        </table>
    );
};

export default Table;
