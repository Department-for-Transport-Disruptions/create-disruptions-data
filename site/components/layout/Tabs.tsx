import kebabCase from "lodash/kebabCase";
import { ReactElement } from "react";

interface TabsProps {
    tabs: Tab[];
    tabsTitle: string;
}

interface Tab {
    tabHeader: string;
    content: ReactElement;
    handleTabClick?: () => void;
}

const Tabs = ({ tabsTitle, tabs }: TabsProps): ReactElement => (
    <div className="govuk-tabs" data-module="govuk-tabs">
        <h2 className="govuk-tabs__title">{tabsTitle}</h2>
        <ul className="govuk-tabs__list">
            {tabs.map((tab, index) => (
                <li
                    key={`tab-${index + 1}`}
                    className={`govuk-tabs__list-item${index === 0 ? " govuk-tabs__list-item--selected" : ""}`}
                >
                    <a className="govuk-tabs__tab" href={`#${kebabCase(tab.tabHeader)}`} onClick={tab.handleTabClick}>
                        {tab.tabHeader}
                    </a>
                </li>
            ))}
        </ul>
        {tabs.map((tab, index) => (
            <div
                key={`tab-panel-${index + 1}`}
                className={`govuk-tabs__panel${index !== 0 ? " govuk-tabs__panel--hidden" : ""}`}
                id={kebabCase(tab.tabHeader)}
            >
                {tab.content}
            </div>
        ))}
    </div>
);

export default Tabs;
