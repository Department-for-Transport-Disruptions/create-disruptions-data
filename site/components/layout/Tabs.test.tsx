import Link from "next/link";
import renderer from "react-test-renderer";
import { describe, expect, it } from "vitest";
import Table from "../form/Table";
import Tabs from "./Tabs";

describe("Tabs", () => {
    it("should render correctly when given a table as the content", () => {
        const tree = renderer
            .create(
                <Tabs
                    tabs={[
                        {
                            tabHeader: "Live",
                            content: (
                                <Table
                                    caption={{ text: "Live disruptions", size: "l" }}
                                    columns={["ID", "Summary", "Affected dates"]}
                                    rows={[
                                        {
                                            header: (
                                                <Link className="govuk-link" href="/dashboard">
                                                    1234
                                                </Link>
                                            ),
                                            cells: ["Test summary", "01/01/2023 onwards"],
                                        },
                                        {
                                            header: (
                                                <Link className="govuk-link" href="/dashboard">
                                                    3211
                                                </Link>
                                            ),
                                            cells: ["Test summary", "02/01/2023 - 02/02-2024"],
                                        },
                                    ]}
                                />
                            ),
                        },
                        {
                            tabHeader: "Upcoming",
                            content: (
                                <Table
                                    caption={{ text: "Live disruptions", size: "l" }}
                                    columns={["ID", "Summary", "Affected dates"]}
                                    rows={[
                                        {
                                            header: (
                                                <Link className="govuk-link" href="/dashboard">
                                                    1234
                                                </Link>
                                            ),
                                            cells: ["Test summary", "01/01/2023 onwards"],
                                        },
                                        {
                                            header: (
                                                <Link className="govuk-link" href="/dashboard">
                                                    3211
                                                </Link>
                                            ),
                                            cells: ["Test summary", "02/01/2023 - 02/02-2024"],
                                        },
                                    ]}
                                />
                            ),
                        },
                    ]}
                    tabsTitle="Disruptions"
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly when given simple text as the content", () => {
        const tree = renderer
            .create(
                <Tabs
                    tabs={[
                        {
                            tabHeader: "Live",
                            content: <h1>Hello</h1>,
                        },
                        {
                            tabHeader: "Upcoming",
                            content: <h1>Test</h1>,
                        },
                    ]}
                    tabsTitle="Disruptions"
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
