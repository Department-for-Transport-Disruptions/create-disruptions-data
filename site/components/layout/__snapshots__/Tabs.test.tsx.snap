// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > should render correctly when given a table as the content 1`] = `
<div
  className="govuk-tabs"
  data-module="govuk-tabs"
>
  <h2
    className="govuk-tabs__title"
  >
    Disruptions
  </h2>
  <ul
    className="govuk-tabs__list"
  >
    <li
      className="govuk-tabs__list-item govuk-tabs__list-item--selected"
    >
      <a
        className="govuk-tabs__tab"
        href="#live"
      >
        Live
      </a>
    </li>
    <li
      className="govuk-tabs__list-item"
    >
      <a
        className="govuk-tabs__tab"
        href="#upcoming"
      >
        Upcoming
      </a>
    </li>
  </ul>
  <div
    className="govuk-tabs__panel"
    id="live"
  >
    <table
      className="govuk-table"
    >
      <caption
        className="govuk-table__caption govuk-table__caption--l"
      >
        Live disruptions
      </caption>
      <thead
        className="govuk-table__head"
      >
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="col"
          >
            ID
          </th>
          <th
            className="govuk-table__header align-middle"
            scope="col"
          >
            Summary
          </th>
          <th
            className="govuk-table__header align-middle"
            scope="col"
          >
            Affected dates
          </th>
        </tr>
      </thead>
      <tbody
        className="govuk-table__body"
      >
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            <a
              className="govuk-link"
              href="/dashboard"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              1234
            </a>
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            Test summary
          </td>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            01/01/2023 onwards
          </td>
        </tr>
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            <a
              className="govuk-link"
              href="/dashboard"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              3211
            </a>
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            Test summary
          </td>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            02/01/2023 - 02/02-2024
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="govuk-tabs__panel govuk-tabs__panel--hidden"
    id="upcoming"
  >
    <table
      className="govuk-table"
    >
      <caption
        className="govuk-table__caption govuk-table__caption--l"
      >
        Live disruptions
      </caption>
      <thead
        className="govuk-table__head"
      >
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="col"
          >
            ID
          </th>
          <th
            className="govuk-table__header align-middle"
            scope="col"
          >
            Summary
          </th>
          <th
            className="govuk-table__header align-middle"
            scope="col"
          >
            Affected dates
          </th>
        </tr>
      </thead>
      <tbody
        className="govuk-table__body"
      >
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            <a
              className="govuk-link"
              href="/dashboard"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              1234
            </a>
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            Test summary
          </td>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            01/01/2023 onwards
          </td>
        </tr>
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            <a
              className="govuk-link"
              href="/dashboard"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              3211
            </a>
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            Test summary
          </td>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            02/01/2023 - 02/02-2024
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Tabs > should render correctly when given simple text as the content 1`] = `
<div
  className="govuk-tabs"
  data-module="govuk-tabs"
>
  <h2
    className="govuk-tabs__title"
  >
    Disruptions
  </h2>
  <ul
    className="govuk-tabs__list"
  >
    <li
      className="govuk-tabs__list-item govuk-tabs__list-item--selected"
    >
      <a
        className="govuk-tabs__tab"
        href="#live"
      >
        Live
      </a>
    </li>
    <li
      className="govuk-tabs__list-item"
    >
      <a
        className="govuk-tabs__tab"
        href="#upcoming"
      >
        Upcoming
      </a>
    </li>
  </ul>
  <div
    className="govuk-tabs__panel"
    id="live"
  >
    <h1>
      Hello
    </h1>
  </div>
  <div
    className="govuk-tabs__panel govuk-tabs__panel--hidden"
    id="upcoming"
  >
    <h1>
      Test
    </h1>
  </div>
</div>
`;
